Add the (basic) ability to persist data to Hyrise and use it in the TPC-H benchmark

phoeinx opened the pull request:
## Main Contribution

This PR enables running the TPC-H Benchmark with mmap-based data.
To use memory-mapping, the `--use_mmap` flag has to be set. 

The rough program flow is as follows:
- Generate tables of `TPCHBenchmark`
- Exchange tables with mmap-based tables
   -  on `Chunk`-Level using `replace_chunk_with_persisted_chunk()` method of `StorageManager`
   - Persist the mmap-based tables to disk
- Run TPC-H Benchmark as usual

When the TPC-H tables already have been persisted:
- Map the saved tables from disk and create tables spanning over the mapped data
- Run TPC-H Benchmark as usual

### Most relevant changes

The most relevant changes happened in `StorageManager`. A good entry point for this PR is the `StorageManager::replace_chunk_with_persisted_chunk()` method. 

The persistence interface is now mainly offered by Table with 
- `Table::persist()` and
- `Table::persist_chunk(ChunkID chunk_id)`.

### Open questions
Most important questions:
For now I would like to focus the discussion on the architecture and interfaces:
- See [comment about OO](https://github.com/hyrise-mp-22-23/hyrise/pull/117#discussion_r1130758699).
-  Table-based persistence interface as described above. Does this make sense?


### Further contributions
This PR also implements:
- type-agnostic mapping of DictionarySegments (previously only DictionarySegments<int> were supported
- mapping of FixedStringDictionarySegments to and from memory region
- new mapping base using `byte *` pointer to avoid having to pad
- rudimentary persistence-file-table handling (being able to write chunks of a table to multiple files and handle references, as well as updating an already existing Chunk file)
- basic testing (which needs to be moved to real tests) -> the `--verify` option also proves correct implementation
- further refactoring

This Pull Request also includes the `storage.json`-PR #99  which enables the cold start of Hyrise.

phoeinx committed:
WIP: Start implementation for basic TPC-H support. See next step in storage_manager.cpp

phoeinx committed:
Implement type-agnostic writing of DictionarySegment.

phoeinx committed:
Fix file index counting up.

phoeinx committed:
Continue implementing mapping DictionarySegments of different data types

phoeinx committed:
Adapt to non-encoded segments, allow multiple data types in DictionarySegments

phoeinx committed:
Adapt StorageManager further to writing and mapping DictionarySegments with different data types

phoeinx committed:
Add padding to uint32_t per mapping part to ensure mappability

phoeinx committed:
Refactor: Rename to calculate_segment_offset_ends to better fit actual operation

phoeinx committed:
Add padding to four_byte align persisted data, <intX_t>DictionaryEncodedSegments work now

phoeinx committed:
Fix alignment calculation, Feat: Map first chunk from TPC-H Benchmark

phoeinx committed:
Add temporary evaluation method for mapped chunks to StorageManager

phoeinx committed:
Fix attribute_vector calculation error. Feat: One chunk mapped correctly now.

phoeinx committed:
Add row-oriented mapped chunk test

phoeinx committed:
Change encoding to FixedStringDictionary for Strings

phoeinx committed:
Implement constructor from memory address for FixedStringVector/Dictionary

phoeinx committed:
Feat: Map chunk with Dictionary and FixedStringDictionaries

phoeinx committed:
Feat: Map chunk with FixedStrings.

phoeinx committed:
Implement table->replace_chunk() method.

phoeinx committed:
Fix overwriting header function.

phoeinx committed:
Feat: Map first whole table of TPC-H.

phoeinx committed:
Feat: Map all Tables of TPC-H Benchmark.

phoeinx committed:
Feat: Get TPC-H 01 to run through.

phoeinx committed:
Add crutches to get TPC-H running until 08.

This depends on our span-based implementation of DictionarySegments etc. We will need to adapt this next.

phoeinx committed:
Revert "Add crutches to get TPC-H running until 08."

This reverts commit 0feb9d2f33d620c2ba2b554a06d4be78e04f52e1.

phoeinx committed:
Refactor DictionarySegment to be span-based by default.

Commit:
Fix Test Cases

phoeinx committed:
Refactor FixedStringDictionarySegment to be span-based by default.

phoeinx committed:
Fix FixedStringSpan implementation (and TPC-H 09)

phoeinx committed:
Fix (provisorial) binary_writer

phoeinx committed:
Implement Table based persistence interface.

phoeinx committed:
Add more explanation to binary_writer_test provisional fix.

phoeinx committed:
Remove map on whole file, just map page-aligned section containing chunk

Commit:
Make Table Dump Again

Commit:
Commeted Out Outdated Test Cases To Make Everything Compile

phoeinx committed:
Adapt read_chunk_header to use already initialized map

phoeinx committed:
Refactor to use std::byte* map of mmaped data.

phoeinx committed:
Refactor mmapped region naming.

phoeinx committed:
Fix big TPC-H benchmarks by removing underflow calculation.

Commit:
Hacky Way To Remove Bins

Commit:
Renamed Stuff and Made Stuff Private

Commit:
Ran Format Skript

Commit:
Deleted Commented Out Method

Commit:
Only One Call to SM to Persist Chunk

Commit:
Makes Stuff Const and to References

Commit:
Moves Stuff to Anonymous Namespace

Commit:
Delete Left Over Binaries

Commit:
Improved Binary Removal

Commit:
Fixed Typo

Commit:
Gave Table Back Control Over Persistion

phoeinx commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -275,110 +439,93 @@ std::ostream& operator<<(std::ostream& stream, const StorageManager& storage_man
+template <typename T>
+void StorageManager::_write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<T>> segment,
+                                                 const std::string& file_name) const {
   /*
```
remove as not up to date anymore

phoeinx commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -275,110 +439,93 @@ std::ostream& operator<<(std::ostream& stream, const StorageManager& storage_man
+  //TODO: What to do with non-compressed AttributeVectors?
+}
+
+void StorageManager::_write_segment_to_disk(const std::shared_ptr<AbstractSegment> abstract_segment,
```
This is worth a discussion of the responsibilities of different classes/objects.

Should the StorageManager handle the whole persisting process and know detailed what to do per SegmentClass?

Or would it be nicer to let each Segment implement a `.byte_representation()` method, which would return a byte array to write out to disk. The Segments already know how to initialize (decode) themselves from a byte array and therefore should be able to encode themselves. 
+ this would keep knowledge about the internal structure in the Segments
+ this would decouple implementation and allow for easy extension

In conclusion, I would strongly argue for the second option.

Bouncner replied:
100% agree.

tjjordan replied:
How do you like the way it is done now?

mweisgut replied:
I agree with the second option and would vote against having the encoding/decoding logic for each segment type in the storage manager.

Alternatively, one could develop a BinarySegmentEncoded/-Decoder that encapsulates the encoding/decoding logic.

phoeinx commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
-  }
+std::pair<uint32_t, uint32_t> StorageManager::_persist_chunk_to_file(const std::shared_ptr<Chunk> chunk,
+                                                                     ChunkID chunk_id,
+                                                                     const std::string& file_name) const {
```
There is quite a bit of code duplication between persisting to an existing file and for a new file here. 
Either resolve code duplication or extract methods?

phoeinx commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
-  // TODO: Remove unneccesary map on whole file
-  const auto* map = reinterpret_cast<uint32_t*>(mmap(NULL, file_bytes, PROT_READ, MAP_PRIVATE, fd, off_t{0}));
-  Assert((map != MAP_FAILED), "Mapping of File Failed.");
+  const auto* persisted_data =
```
consider renaming to `persisted_chunk` ?

phoeinx commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -390,141 +537,204 @@ void StorageManager::write_chunk_to_disk(const std::shared_ptr<Chunk>& chunk,
-    segments.emplace_back(std::make_shared<DictionarySegment<int32_t>>(map + segment_offset_end / 4));
+    resolve_data_type(column_definitions[segment_index], [&](auto type) {
+      using ColumnDataType = typename decltype(type)::type;
+      if constexpr (std::is_same<ColumnDataType, pmr_string>::value) {
```
Add comment that we can only persist `FixedStringDictionarySegments` and therefore know that this is one.

Bouncner commented:
> Table-based persistence interface as described above. Does this make sense?

I guess the table itself does not write. Without looking at the code, what does `table->persist()` do?

Commit:
Small Refactorings for OOP

Commit:
DictSegment Serialzies Itself Now

Commit:
FixedStringDictSegment Serializes Itself Now

Commit:
WIP, does not compile

Commit:
Removed BitPacking Support, compiles now

Commit:
Cleanup

Commit:
Followup

Commit:
Some Cleanup

Commit:
More OOP

Commit:
Even More OOP

tjjordan commented on src/benchmarklib/abstract_table_generator.cpp:
```diff
@@ -342,11 +342,50 @@ void AbstractTableGenerator::generate_and_store() {
+  _table_info_by_name = table_info_by_name;
+
+  // Persist tables
+  // As last step after completing encoding etc.
```
It this comment outdated?

Commit:
Some Commits and Some Cleanup

mweisgut commented on src/lib/storage/table.hpp:
```diff
@@ -225,7 +227,16 @@ class Table : private Noncopyable {
+   * TODO: Explanation.
+   */
+
+  void persist();
```
Related to Martin's comment about not being responsible for persisting a table in the `Table` class: `persist_table` in the storage manager might be an alternative.

mweisgut commented on src/lib/storage/table.hpp:
```diff
@@ -225,7 +227,16 @@ class Table : private Noncopyable {
+  void persist();
+
  protected:
+  void _persist_chunk(ChunkID chunk_id);
```
Do we have a use case that requires any component to persist/snapshot an individual chunk? Assuming we persist a table at time t1 and later at t2 again since the table has more chunks at t2, we want to have the functionality of persisting individual chunks that persist the delta instead of writing all chunks again. However, this could also be implemented as an optimization of `persist_table`. Then, we would call `persist_table` at t1 and t2, but at the second call, only the delta of immutable chunks will be persisted.

tjjordan replied:
For better object orientation, I think the chunks should be able to persist themselves.

Bouncner replied:
I wouldn't argue that this is better object orientation. Apart from that: OO is soooooo 2000s! :D

Commit:
Some Comments

Commit:
copy_using_allocator() Supportes std::spans Now

tjjordan commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -137,8 +130,9 @@ ChunkOffset DictionarySegment<T>::size() const {
 template <typename T>
 std::shared_ptr<AbstractSegment> DictionarySegment<T>::copy_using_allocator(
     const PolymorphicAllocator<size_t>& alloc) const {
+  Assert(_dictionary_base_vector, "Cannot copy based on span-only DictionarySegment.");
```
We support it now :)
(Currently on the branch some_feedback)

tjjordan commented on src/lib/storage/table.cpp:
```diff
@@ -450,4 +455,16 @@ size_t Table::memory_usage(const MemoryUsageCalculationMode mode) const {
+
+void Table::_persist_chunk(ChunkID chunk_id) {
+  auto& storage_manager = Hyrise::get().storage_manager;
+  storage_manager.replace_chunk_with_persisted_chunk(_chunks[chunk_id], chunk_id, this);
```
Somewhere we need to check if the chunk is immutable.

tjeyy committed:
Make It Compile Without UMAP

janlemcke committed:
storage.json

tjeyy committed:
Merge branch 'some_feedback' into feature/enable_running_tpch_benchmark

tjeyy committed:
Resolved Merge Conflicts

tjjordan commented on src/lib/storage/storage_manager.cpp:
```diff
@@ -75,6 +230,15 @@ std::shared_ptr<Table> StorageManager::get_table(const std::string& name) const
+      return name;
+    }
+  }
+  return std::string{};
```
Maybe no.

tjeyy committed:
Adapted Some Feedback

Commit:
Refactorings

Commit:
Fixed use mmap flag

Commit:
Fix

Commit:
Added Integration Test

Bouncner commented on src/benchmark/tpch_benchmark.cpp:
```diff
@@ -50,13 +50,15 @@ int main(int argc, char* argv[]) {
                    "and 'orders' are sorted by 'l_shipdate' and 'o_orderdate' for improved chunk pruning. Both are "
-                   "legal TPC-H input data.", cxxopts::value<std::string>()->default_value("None")); // NOLINT
+                   "legal TPC-H input data.", cxxopts::value<std::string>()->default_value("None")) // NOLINT
+    ("use_mmap", "Use storage.json to load TPCH data using the fileformat for chunks that hold their data with mmap."); // NOLINT
```
Site note: the user of the benchmarks/Hyrise should not care/be aware about the name of internal files that are used for persistence.

Bouncner commented on src/benchmarklib/abstract_table_generator.cpp:
```diff
@@ -342,11 +343,41 @@ void AbstractTableGenerator::generate_and_store() {
   }
 
+  /**
+   * Persist the tables in binary files if we want to use the new file format
```
Not of relevance: why not use the entire width of 120 chars for the comment?

Bouncner commented on src/benchmarklib/abstract_table_generator.cpp:
```diff
@@ -407,4 +438,52 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
+    const auto column_definitions = storage_manager.get_table_column_definitions_from_json(table_name);
+
+    // Load chunks from files
+    for (auto index = size_t{0}; index < file_count + 1; ++index) {
```
Since the `+ 1` is not immediately clear (imo), I'd add a comment why you're using it here.

Bouncner commented on src/benchmarklib/abstract_table_generator.cpp:
```diff
@@ -407,4 +438,52 @@ std::unordered_map<std::string, BenchmarkTableInfo> AbstractTableGenerator::_loa
+    const auto persistence_data = mapping.second;
+    auto file_name = persistence_data.file_name;
+    const auto file_count = persistence_data.file_index;
+    auto total_chunks = std::vector<std::shared_ptr<Chunk>>{};
```
strange name

Bouncner replied:
Could be reserved().

Bouncner commented on src/benchmarklib/encoding_config.cpp:
```diff
@@ -4,7 +4,13 @@
 namespace hyrise {
 
-EncodingConfig::EncodingConfig() : EncodingConfig{SegmentEncodingSpec{EncodingType::Dictionary}} {}
+// Because it is hardly possible to save regular std:strings on disk, we use FixedStringDictionary as the default
```
It's absolutely possible to persist them on disk, but doesn't help us here and needs some magic allocator stuff.

"hardly possible" is a strange wording. :)

(Not to be addressed)

Bouncner commented on src/benchmarklib/tpch/tpch_table_generator.cpp:
```diff
@@ -125,9 +126,24 @@ std::unordered_map<std::string, BenchmarkTableInfo> TPCHTableGenerator::generate
+  auto cache_directory = std::string{};
+
+  if (_benchmark_config->use_mmap) {
+    if (!std::filesystem::is_directory("tpch_cached_tables_storage_json")) {
```
Not to be addressed: why use a flag `mmap` instead of simply passing a directory name that automatically states if we want to use persistency and which would avoid such hard-coded names?

Bouncner commented on src/lib/statistics/generate_pruning_statistics.cpp:
```diff
@@ -63,7 +63,10 @@ void generate_chunk_pruning_statistics(const std::shared_ptr<Chunk>& chunk) {
+        // DictionarySements use std::spans now, but this method still requires a std::vector. Because of this, we
+        // create a vector with the content of the span.
+        pmr_vector<ColumnDataType> dictionary_vector{dictionary.begin(), dictionary.end()};
+        create_pruning_statistics_for_segment(*segment_statistics, dictionary_vector);
```
What this part of the cold start measurements?

Bouncner commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -30,76 +31,76 @@ template <typename T>
-        break;
-      }
+  const auto header_data = reinterpret_cast<const uint32_t*>(start_address);
+  const auto encoding_type = PersistedSegmentEncodingType{header_data[ENCODING_TYPE_OFFSET_INDEX]};
```
Do we need this additional enum? All it stores has been available in  existing enums before?

Bouncner commented on src/lib/storage/dictionary_segment.cpp:
```diff
@@ -209,7 +207,26 @@ std::shared_ptr<const BaseCompressedVector> DictionarySegment<T>::attribute_vect
+  StorageManager::export_value(static_cast<uint32_t>(compressed_vector_type_id), ofstream);
+
+  // Ee need to ensure that every part can be mapped with a uint32_t map.
+  StorageManager::export_value(static_cast<uint32_t>(dictionary()->size()), ofstream);
```
Not to be addressed: those methods should not be part of the stroage manager but rather some util class.

Bouncner commented on src/lib/storage/dictionary_segment.hpp:
```diff
@@ -94,15 +91,15 @@ class DictionarySegment : public BaseDictionarySegment {
  protected:
-  const std::shared_ptr<const pmr_vector<T>> _dictionary;
-  std::shared_ptr<const std::span<const T>> _dictionary_span;
+  const std::shared_ptr<const pmr_vector<T>> _dictionary_base_vector;
```
I am fine with it.

Bouncner commented on src/lib/storage/fixed_string_dictionary_segment.cpp:
```diff
@@ -16,11 +17,75 @@ FixedStringDictionarySegment<T>::FixedStringDictionarySegment(
+      break;
+    }
+    case PersistedSegmentEncodingType::DictionaryEncoding16Bit: {
+      auto* const attribute_vector_address =
```
Not to be addressed: There is a lot of duplication here and in the Dictionary class. That could  have been put into little helpers. There are also resolve patterns that would have helped to get rid of a lot of code here.

Bouncner commented on src/lib/storage/fixed_string_dictionary_segment.hpp:
```diff
@@ -70,13 +72,23 @@ class FixedStringDictionarySegment : public BaseDictionarySegment {
+
+ private:
+  // Constants used for the construction of FixedStringDictionarySegments from memory-mapped storage.
+  static constexpr auto ENCODING_TYPE_OFFSET_INDEX = uint32_t{0};
```
Much much duplication duplication.

Bouncner reviewed (commented):
There is nothing that needs to be addressed this week. Just some remarks that might be of help in general or helps for paper discussions. Haven't looked at the storage manager yet.

janlemcke committed:
Added whisker plot for hyperfine results.

phoeinx committed:
Benchmark scripts for MemoryFootprint and TPC-H with/without mmap performance (#122)

* Added initial setup for CGroups

* memory.max set & TPCH as default executing target.

* Adapt scripts for easier parameter passing.

* Fix variable renaming bug.

* Fix printout.

* Adapt to lower size benchmark, add output

* Fix output format

* Hot fix for binary file deleting after benchmark

* Hot fix to fix output file parameter being added correctly

* Add .bin files to gitignore, write new python script for cgroup benchmarking.

* Adapt cgroups benchmarking script.

* Fix execution directory.

* Set memory.max to unlimited first.

* Wild try to make memory limitation work.

* Add memory stat data printout to better understand reclaiming.

* Adapt to correct printing format.

* Add assertion for unmapping.

* Further tries with unmapping.

* Start benchmark immediately in memory limited cgroup

* Fix actually setting high limit

* Build without jemalloc

* Add measure memory peak script.

* Add benchmark realistic scenario mmap

* Revert "Build without jemalloc"

This reverts commit 622363baf93381ef53932bccc8c732bd894700ec.

* Fix for not deleted binaries.

* Add multicore benchmark script for hyrise master.

* Edited Skript

* Works Now

* Add further analyze scripts

* Make memory limit script runnable

* Fix script.

* Fix script.

* Finish Multi Core script.

* Finish visualize limited memory script.

* Update scripts.

* Set realistic scenario benchmarks up with parameters for final eval.

* Change warmup to 5 seconds to reduce overall runtime

* Setup cgroup limited benchmark script for final measurements.

* Setup big final benchmark for presentation

* Fix command.

* Fix hyrise multicore benchmark script.

* Don't wait after subprocess timeout

* Handle Timeout.

* Ok, now the subprocesses should work.

* More changes to benchmark scripts.

* Also kill child processes.

* Fix terminating and paths.

* Set final countdown/timout.

* Adapt for visualization for presentation.

* Adapt scripts for report benchmarking.

* Setup test run

* Make cgroup measurements single-threaded.

* Revert "Setup test run"

This reverts commit 7797fbc566e11f8f06ae1a93353c62e249bb285d.

* Fix writing different benchmarking results to same file.

* Fix bug in fixed cores benchmark scripts.

* Adapt to run of fixed core benchmarks only.

* Add visualization for fixed core benchmarks.

* Setup reproducibility benchmark for underperforming hyrise master

* Setup reproducibility benchmark for memory limited mmap

* Adapt benchmark script for reproducibility benchmarks.

* Refactor benchmarking and visualization scripts.

Refactorings include:
- New directory structure to seperate mp-2023 scripts from previous scripts.
- Adding comments to each script explaining what it does.
- Some minor clean up, like e.g. using os.path.join() instead of formatting strings to create a new path.

As most of these scripts are highly specific for the MP, and will most likely only be used as reference (if at all), the current clean up and documentation state looks alright to me.

* Add analyze_memory_peaks script to mp-2023 scripts.

* Fix rebase occurred changes.

---------

Co-authored-by: Robert Richter <robert.richter@rrcomtech.com>
Co-authored-by: Tobias Jordan <tjjordan@githb.nomail>

phoeinx committed:
Benchmark/pagefaults (#128)

Add Pagefault Measurements and clean-up.

---------

Co-authored-by: Sandro Sp <38314662+SanJSp@users.noreply.github.com>

phoeinx committed:
Comment CMakeLists executable folders back in.

phoeinx committed:
Run Format Script.

phoeinx committed:
Remove completely outcommented util file.

Commit:
Make it buld

Commit:
Removed Outdated Tests

Commit:
Removed Persistence Capability From Tables

Commit:
Merge branch 'master' into feature/enable_running_tpch_benchmark
